compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt

```diff
@@ -0,0 +1,348 @@
+        } else if (call.type.isShort()) {
+            IrConstImpl.short(call.startOffset, call.endOffset, call.type, v as Short)
+        } else if (call.type.isLong()) {
+            IrConstImpl.long(call.startOffset, call.endOffset, call.type, v as Long)
+        } else if (call.type.isDouble()) {
+            IrConstImpl.double(call.startOffset, call.endOffset, call.type, v as Double)
+        } else if (call.type.isFloat()) {
+            IrConstImpl.float(call.startOffset, call.endOffset, call.type, v as Float)
+        } else if (call.type.isString()) {
+            IrConstImpl.string(call.startOffset, call.endOffset, call.type, v as String)
+        } else {
+            throw IllegalArgumentException("Unexpected IrCall return type")
+        }
+    }
+
+    private fun tryFoldingUnaryOps(call: IrCall): IrExpression {
+        val owner = call.dispatchReceiver
+        if (owner !is IrConst<*>)
+            return call
+
+        val evaluator = UNARY_OP_TO_EVALUATOR[UnaryOp.get(call.descriptor)] ?: return call
+
+        return buildIrConstant(call, evaluator(owner.value!!))
+    }
+
+    // Do type promotion so that we don't have to hard-code all the table.
+    // T.{plus, minus, times, rem, mod, div, compareTo}(T) where T is {Int, Long, Float, Double} are defined.
+    // A map without type promotion would be 7 (operators) * 36 (# number types ^ 2) = 252 lines of lambda.
+    private fun promotePrimitive(a: Any, b: Any): Pair<Any, Any> {
+        if (a is Number && b is Number) {
+            if (a is Double || b is Double)
+                return Pair(a.toDouble(), b.toDouble())
+
+            if (a is Float || b is Float)
+                return Pair(a.toFloat(), b.toFloat())
+
+            if (a is Long || b is Long)
+                return Pair(a.toLong(), b.toLong())
+
+            return Pair(a.toInt(), b.toInt())
+        }
+
+        return Pair(a, b)
+    }
+
+    private fun tryFoldingBinaryOps(call: IrCall): IrExpression {
+        val lhs = call.dispatchReceiver
+        val rhs = call.getValueArgument(0)
+
+        if (lhs !is IrConst<*> || rhs !is IrConst<*>)
+            return call
+
+        val evaluator = BINARY_OP_TO_EVALUATOR[BinaryOp.get(call.descriptor, rhs.kind)] ?: return call
+
+        val (v1, v2) = promotePrimitive(lhs.value!!, rhs.value!!)
+        val evaluated = try {
+            evaluator(v1, v2)
+        } catch (e: ArithmeticException) {
+            // Don't cast a runtime exception into compile time. E.g., division by zero.
+            return call
+        }
+
+        return buildIrConstant(call, evaluated)
+    }
+
+    override fun lower(irFile: IrFile) {
+        irFile.transformChildrenVoid(object : IrElementTransformerVoid() {
+            override fun visitCall(expression: IrCall): IrExpression {
+                expression.transformChildrenVoid(this)
+                return when (expression.valueArgumentsCount) {
```

Comment:
We have 2 representations for compare and equals operation in IR:
1 < 1 is binary operation with 2 argument 
1.compareTo(1) - call with dispatch receiver and 1 argument 
Maybe call.receiverAndArgs()  would be worth
Seems this cases are not tested.

Please also note that:
-0.0 < 0.0 calculation is performed in IEEE754 arithmetics and
-0.0.compareTo(0.0) in non-IEEE754 
This affects  registerSymmetricBinaryOp<Double, Double>(DOUBLE, opType, "compareTo") { a, b -> a.compareTo(b) } same applies for == and equals operation. (Similar with Floats)

Seems other operations should be equals to each other (+ and plus and so on), but maybe I miss smth @dnpetrov

Reply:
Thanks for the pointers! I've been wondering where those IrBuiltins are used :) Please take a look at the new commit, where the support of folding some IrBuiltins, mainly for those comparisons, are added.